src/metabase/models/serialization.clj

```diff
@@ -1482,6 +1509,65 @@
   (set/union (viz-click-behavior-descendants  viz)
              (viz-column-settings-descendants viz)))
 
+;;; Common transformers
+
+(defn fk "Export Foreign Key" [model & [field-name]]
+  (cond
+    ;; this `::fk` is used in tests to determine that foreign keys are handled
+    (= model :model/User)  {::fk true :export *export-user* :import *import-user*}
+    (= model :model/Table) {::fk true :export *export-table-fk* :import *import-table-fk*}
+    (= model :model/Field) {::fk true :export *export-field-fk* :import *import-field-fk*}
+    field-name             {::fk    true
+                            :export #(*export-fk-keyed* % model field-name)
+                            :import #(*import-fk-keyed* % model field-name)}
+    :else                  {::fk true :export #(*export-fk* % model) :import #(*import-fk* % model)}))
+
+(defn nested "Nested entities" [model backward-fk opts]
+  (let [model-name (name model)
+        sorter     (:sort-by opts :created_at)
+        key-field  (:key-field opts :entity_id)]
+    {::nested     true
+     :model       model
+     :backward-fk backward-fk
+     :export      (fn [data]
+                    ;; if this looks weird, see :series hydration for DashboardCard, it supplies Cards while we need
+                    ;; to serialize DashboardCardSeries instances
+                    (when-not (every? #(t2/instance-of? model %) data)
+                      (assert false
+                              (format "Nested data is expected to be a %s, not %s" model (t2/model (first data)))))
+                    ;; `nil? data` check is for `extract-one` case in tests; make sure to add empty vectors in
+                    ;; `extract-query` implementations for nested collections
+                    (try
+                      (->> (or data (when (nil? data)
+                                      (t2/select model backward-fk (:id *current*))))
```

Comment:
Virtually all tests in `v2.e2e-tests` will fail and I just did not want to introduce that much churn :)

Reply:
fair enough. would be nice to clean up in a follow up PR though? I count only 12 uses of `extract-one` in `v2.e2e-tests` so it might not be that much work to replace them with `extract-all` calls